File: sdk/ts/src/streamEventMeta.ts
import type { SSEEvent } from './types';
import { firstPositiveNumber, firstString } from './ordering';

export interface ConversationLifecyclePatch {
    running: boolean;
    stage: string;
    status: string;
}

export function eventSequenceValue(event?: SSEEvent, fallback = 1): number {
    return firstPositiveNumber(event?.pageIndex, event?.iteration, event?.eventSeq, fallback);
}

export function eventIterationValue(event?: SSEEvent, fallback = 0): number {
    return firstPositiveNumber(event?.iteration, event?.pageIndex, fallback);
}

export function terminalStatusForType(type = ''): string {
    const normalized = String(type || '').trim().toLowerCase();
    if (normalized === 'turn_failed') return 'failed';
    if (normalized === 'turn_canceled') return 'canceled';
    return 'completed';
}

export function conversationStatusForStreamEvent(type = '', status = '', fallback = 'running'): string {
    const explicit = String(status || '').trim().toLowerCase();
    if (explicit === 'succeeded' || explicit === 'success' || explicit === 'completed' || explicit === 'done') {
        return 'succeeded';
    }
    if (explicit === 'failed' || explicit === 'error') return 'failed';
    if (explicit === 'canceled' || explicit === 'cancelled') return 'canceled';
    if (explicit) return explicit;
    const normalizedType = String(type || '').trim().toLowerCase();
    if (normalizedType === 'turn_failed') return 'failed';
    if (normalizedType === 'turn_canceled') return 'canceled';
    if (normalizedType === 'turn_completed') return 'succeeded';
    return String(fallback || 'running').trim() || 'running';
}

export function conversationStageForStreamEvent(type = '', fallback = 'executing'): string {
    const normalized = String(type || '').trim().toLowerCase();
    if (normalized === 'turn_failed') return 'error';
    if (normalized === 'turn_canceled') return 'canceled';
    if (normalized === 'turn_completed') return 'done';
    return String(fallback || 'executing').trim() || 'executing';
}

export function conversationLifecyclePatchForStreamPhase(
    phase = '',
    payload: Partial<SSEEvent> | Record<string, unknown> = {}
): ConversationLifecyclePatch | null {
    const normalizedPhase = String(phase || '').trim().toLowerCase();
    if (normalizedPhase === 'thinking') {
        return {
            running: true,
            stage: 'thinking',
            status: conversationStatusForStreamEvent(String(payload?.type || ''), String(payload?.status || ''), 'running')
        };
    }
    if (normalizedPhase === 'executing') {
        return {
            running: true,
            stage: 'executing',
            status: conversationStatusForStreamEvent(String(payload?.type || ''), String(payload?.status || ''), 'running')
        };
    }
    if (normalizedPhase === 'eliciting') {
        return {
            running: true,
            stage: 'eliciting',
            status: conversationStatusForStreamEvent(String(payload?.type || ''), String(payload?.status || ''), 'pending')
        };
    }
    if (normalizedPhase === 'terminal') {
        return {
            running: false,
            stage: conversationStageForStreamEvent(String(payload?.type || ''), 'done'),
            status: conversationStatusForStreamEvent(String(payload?.type || ''), String(payload?.status || ''), 'succeeded')
        };
    }
    return null;
}

export function modelStepStatusForEvent(event: SSEEvent, existingStatus = '', fallbackStatus = 'running') {
    const explicitStatus = firstString(event?.status);
    if (explicitStatus) return explicitStatus;
    const type = firstString(event?.type).toLowerCase();
    if (type === 'model_completed') return 'completed';
    if (type === 'text_delta') return 'streaming';
    return firstString(fallbackStatus, existingStatus, 'running');
}

export function executionGroupStatusForEvent(event: SSEEvent, existingStatus = '', fallbackStatus = 'running') {
    const explicitStatus = firstString(event?.status);
    if (explicitStatus) return explicitStatus;
    const type = firstString(event?.type).toLowerCase();
    if (type === 'model_completed') return 'completed';
    if (type === 'text_delta') {
        const normalized = firstString(existingStatus).toLowerCase();
        if (['completed', 'done', 'success', 'succeeded', 'failed', 'error', 'canceled', 'cancelled', 'terminated'].includes(normalized)) {
            return existingStatus;
        }
        return 'streaming';
    }
    return firstString(fallbackStatus, existingStatus, 'running');
}

File: sdk/stream_event_meta.go
package sdk

import (
	"strings"
	"time"

	"github.com/viant/agently-core/runtime/streaming"
)

func terminalStatusForEventType(t streaming.EventType) TurnStatus {
	switch t {
	case streaming.EventTypeTurnFailed:
		return TurnStatusFailed
	case streaming.EventTypeTurnCanceled:
		return TurnStatusCanceled
	default:
		return TurnStatusCompleted
	}
}

func turnStatusFromString(status string, fallback TurnStatus) TurnStatus {
	switch strings.ToLower(strings.TrimSpace(status)) {
	case "queued":
		return TurnStatusQueued
	case "running":
		return TurnStatusRunning
	case "waiting_for_user":
		return TurnStatusWaitingForUser
	case "completed", "succeeded", "success", "done":
		return TurnStatusCompleted
	case "failed", "error":
		return TurnStatusFailed
	case "canceled", "cancelled", "terminated":
		return TurnStatusCanceled
	default:
		if fallback != "" {
			return fallback
		}
		return TurnStatusCompleted
	}
}

func elicitationStatusForString(status string, fallback ElicitationStatus) ElicitationStatus {
	switch strings.ToLower(strings.TrimSpace(status)) {
	case "pending":
		return ElicitationStatusPending
	case "accepted":
		return ElicitationStatusAccepted
	case "declined", "rejected":
		return ElicitationStatusDeclined
	case "canceled", "cancelled":
		return ElicitationStatusCanceled
	default:
		if fallback != "" {
			return fallback
		}
		return ElicitationStatusDeclined
	}
}

func stepStatusFromString(status, fallback string) string {
	return firstNonEmptyString(status, fallback)
}

func modelStepStatusForEvent(event *streaming.Event, existingStatus, fallbackStatus string) string {
	if event == nil {
		return firstNonEmptyString(fallbackStatus, existingStatus, string(TurnStatusRunning))
	}
	if status := strings.TrimSpace(event.Status); status != "" {
		return status
	}
	if event.Type == streaming.EventTypeTextDelta {
		return "streaming"
	}
	return firstNonEmptyString(fallbackStatus, existingStatus, string(TurnStatusRunning))
}

func completedAtForEvent(event *streaming.Event) *time.Time {
	if event == nil {
		return nil
	}
	if event.CompletedAt != nil {
		return event.CompletedAt
	}
	t := event.CreatedAt
	return &t
}

func elicitationStatusForEventStatus(status string) ElicitationStatus {
	return elicitationStatusForString(status, ElicitationStatusDeclined)
}

func firstNonEmptyString(values ...string) string {
	for _, value := range values {
		if trimmed := strings.TrimSpace(value); trimmed != "" {
			return trimmed
		}
	}
	return ""
}
